### PR TITLE
Do not copy results of reductions to host

### PR DESCRIPTION
**Pull request by kaushikcfd**

### CODE REVIEW

File: meshmode/array_context.py [hunk 1]
```diff
@@ -339,16 +339,15 @@ def where_inner(inner_crit, inner_then, inner_else):
...
+        return cl_array.sum(a, dtype=dtype, queue=self._array_context.queue)
 
     def min(self, a):
         import pyopencl.array as cl_array
-        return cl_array.min(a, queue=self._array_context.queue).get()[()]
+        return cl_array.min(a, queue=self._array_context.queue)
 
     def max(self, a):
         import pyopencl.array as cl_array
-        return cl_array.max(a, queue=self._array_context.queue).get()[()]
```
Comment (inducer) [inline meshmode/array_context.py:351]: This is a pretty big compatibility break. Want to elaborate why this is necessary?

Comment (kaushikcfd) [reply meshmode/array_context.py:351]: 1. Forces the lazy code path to eagerly compute, as there is an implicit freeze here.
2. The implicit device->host transfers might become evident if in the future we also support `axis`.

Comment (kaushikcfd) [reply meshmode/array_context.py:351]: This is breaking a lot of tests :astonished:. Maybe for the special case when the shape of the array to be returned is (), we can just return the value.

Comment (inducer) [reply meshmode/array_context.py:351]: You know, I'm coming around to this. Copying reduction results *is* inconsistent, and forcing a lazy freeze is not desirable. I suspect we just need to bite the bullet and make this incompatible change. It's not great, but the current state is arguably just incorrect.

File: meshmode/array_context.py [hunk 2]
```diff
@@ -339,16 +339,15 @@ def where_inner(inner_crit, inner_then, inner_else):
 
     def sum(self, a, dtype=None):
```
Comment (inducer) [inline meshmode/array_context.py:340]: `la.norm` probably needs the same treatment.

File: meshmode/array_context.py (hunk 1 above)
Comment (inducer) [inline meshmode/array_context.py:351]: We *could* do a backward compatibility job here with a mode switch, but I don't think that's a good idea. Do you agree?

File: meshmode/array_context.py (hunk 2 above)
Comment (kaushikcfd) [reply meshmode/array_context.py:379]: la.norm in meshmode is realized as calls to actx.np.linalg.(sum|max). So the current implemenation should be fine.

File: meshmode/array_context.py (hunk 1 above)
Comment (kaushikcfd) [reply meshmode/array_context.py:390]: Difficult to make the claim that breaking back-compatibility is a good idea. But, it was broken when compared to other array ops provided by ArrayContext.

**Comment by kaushikcfd** [comment]
@inducer: Another change that had to be done: actx.np.linalg.norm of DOFArray just returns a DOFArray of the thawed norms. I have un-deprecated `flat_norm` that still takes a DOFArray and returns the norm of the frozen norm of it's sub arrays. 

(The tests are failing, as this breaks back-compat. If there's no objection to this direction, I can open PRs to apply the necessary changes to the downstream projects.)

File: meshmode/array_context.py (hunk 2 above)
Comment (inducer) [reply meshmode/array_context.py:379]: OIC.

:+1:

**Comment by inducer** [comment]
> actx.np.linalg.norm of DOFArray just returns a DOFArray of the thawed norms.

I don't understand. Is that the state you would like, or the current state, or...? It doesn't seem desirable, unless I'm missing something.

**Comment by kaushikcfd** [comment]
Closing, as it was handled by https://github.com/inducer/arraycontext/pull/43.

**Pull request closed without merging**